packages/server/src/fhir/repo.ts

```diff
@@ -557,6 +565,63 @@ export class Repository extends BaseRepository implements FhirRepository {
     }
   }
 
+  private async validateProfiles(resource: Resource, profileURLs: string[]): Promise<void> {
+    for (const url of profileURLs) {
+      const loadStart = process.hrtime.bigint();
+      const profile = await this.loadProfile(url);
+      const loadTime = Number(process.hrtime.bigint() - loadStart);
+      if (!profile) {
+        logger.warn(`Unknown profile referenced in ${resource.resourceType}/${resource.id}: ${url}`);
+        continue;
+      }
+      const validateStart = process.hrtime.bigint();
+      validate(resource, profile);
+      const validateTime = Number(process.hrtime.bigint() - validateStart);
+      logger.debug(
+        `Profile validation timing: load=${formatDuration(loadTime)}; validate=${formatDuration(validateTime)}`
+      );
+    }
+  }
+
+  private async loadProfile(url: string): Promise<StructureDefinition | undefined> {
+    const redis = getRedis();
+    const cacheKey = `Project/${this.context.project}/StructureDefinition/${url}`;
+    // Try retrieving from cache
+    const cachedProfile = await redis.get(cacheKey);
+    if (cachedProfile) {
+      return (JSON.parse(cachedProfile) as CacheEntry<StructureDefinition>).resource;
+    }
+
+    // Fall back to loading from the DB; descending version sort approximates version resolution for some cases
+    const profile = await this.searchOne<StructureDefinition>({
+      resourceType: 'StructureDefinition',
+      filters: [
+        {
+          code: 'url',
+          operator: FhirOperator.EQUALS,
+          value: url,
+        },
+      ],
+      sortRules: [
+        {
+          code: 'version',
+          descending: true,
+        },
+      ],
+    });
+
+    if (profile) {
+      // Store loaded profile in cache
+      await redis.set(
+        cacheKey,
+        JSON.stringify({ resource: profile, projectId: profile.meta?.project }),
```

Comment:
Is there a reason we need `projectId` in this object, considering it's already in the Redis cache key?  Doesn't look like we're using it.

Reply:
I had added it just for consistency with the other objects stored in cache, but I'm cool with removing it if it's not necessary to maintain that parity